diff --git a/target/linux/mediatek/files-6.6/drivers/net/ethernet/mediatek/mtk_hnat/hnat_nf_hook.c b/target/linux/mediatek/files-6.6/drivers/net/ethernet/mediatek/mtk_hnat/hnat_nf_hook.c
--- a/target/linux/mediatek/files-6.6/drivers/net/ethernet/mediatek/mtk_hnat/hnat_nf_hook.c
+++ b/target/linux/mediatek/files-6.6/drivers/net/ethernet/mediatek/mtk_hnat/hnat_nf_hook.c
@@ -1002,7 +1002,7 @@ static unsigned int is_ppe_support_type(struct sk_buff *skb)
 }
 
 static unsigned int do_hnat_cpu_to_ge(struct sk_buff *skb)
 {
-	if (unlikely(skb_shinfo(skb)->frag_list))
+	if (unlikely(skb_is_gso(skb) || skb_shinfo(skb)->frag_list))
                 return -1;
 	if (unlikely(skb->mark == HNAT_EXCEPTION_TAG))
 				return -1;
