diff --git a/package/mtk/applications/mtwifi-cfg/files/mtwifi.sh b/package/mtk/applications/mtwifi-cfg/files/mtwifi.sh
--- a/package/mtk/applications/mtwifi-cfg/files/mtwifi.sh
+++ b/package/mtk/applications/mtwifi-cfg/files/mtwifi.sh
@@ -7,7 +7,8 @@ append DRIVERS "mtwifi"
 
 detect_mtwifi() {
 	local idx ifname
-	local band hwmode htmode htbsscoex ssid dbdc_main channel
+	local band hwmode htmode htbsscoex ssid dbdc_main channel key
+	key="immortalwrt"
 	if [ -d "/sys/module/mt_wifi" ]; then
 		dev_list="$(l1util list)"
 		for dev in $dev_list; do
@@ -55,7 +56,8 @@ detect_mtwifi() {
 					set wireless.default_${dev}.network=lan
 					set wireless.default_${dev}.mode=ap
 					set wireless.default_${dev}.ssid=${ssid}
-					set wireless.default_${dev}.encryption=none
+					set wireless.default_${dev}.encryption=psk2
+					set wireless.default_${dev}.key=${key}
 EOF
 				uci -q commit wireless
 			}
