diff --git a/base/include/osal.h b/base/include/osal.h
--- a/base/include/osal.h	2023-12-29 05:52:41.000000000 +0000
+++ b/base/include/osal.h	2026-07-20 13:04:40.613966280 +0000
@@ -244,6 +244,13 @@
 	unsigned long usec;
 };
 
+typedef long __kernel_time_t;
+
+struct timeval {
+	__kernel_time_t		tv_sec;		/* seconds */
+	__kernel_suseconds_t	tv_usec;	/* microseconds */
+};
+
 struct osal_op_history {
 	struct ring ring_buffer;
 	struct osal_op_history_entry *queue;
diff --git a/platform/consys_hw.c b/platform/consys_hw.c
--- a/platform/consys_hw.c	2023-12-29 05:52:41.000000000 +0000
+++ b/platform/consys_hw.c	2026-07-20 13:04:40.617299486 +0000
@@ -22,6 +22,7 @@
 #include <linux/delay.h>
 #include <linux/of_device.h>
 #include <linux/of_reserved_mem.h>
+#include <linux/clk.h>
 
 #include "osal.h"
 #include "consys_hw.h"
@@ -518,6 +519,7 @@
 int mtk_conninfra_probe(struct platform_device *pdev)
 {
 	int ret = -1;
+	struct clk *mcu_clk, *ap_conn_clk;
 
 	if (pdev)
 		g_pdev = pdev;
@@ -532,6 +534,18 @@
 		return -2;
 	}
 
+	mcu_clk = devm_clk_get(&pdev->dev, "mcu");
+	if (IS_ERR(mcu_clk))
+		pr_err("mcu clock not found\n");
+	else if (clk_prepare_enable(mcu_clk))
+		pr_err("mcu clock configuration failed\n");
+
+	ap_conn_clk = devm_clk_get(&pdev->dev, "ap2conn");
+	if (IS_ERR(ap_conn_clk))
+		pr_err("ap2conn clock not found\n");
+	else if (clk_prepare_enable(ap_conn_clk))
+		pr_err("ap2conn clock configuration failed\n");
+
 	if (consys_hw_ops == NULL)
 		consys_hw_ops = (const struct consys_hw_ops_struct*)g_conninfra_plat_data->hw_ops;
 	if (consys_hw_ops == NULL) {
@@ -660,4 +674,3 @@
 	g_conninfra_dev_cb = NULL;
 	return 0;
 }
-
diff --git a/platform/consys_reg_mng.c b/platform/consys_reg_mng.c
--- a/platform/consys_reg_mng.c	2023-12-29 05:52:41.000000000 +0000
+++ b/platform/consys_reg_mng.c	2026-07-20 13:04:40.617299486 +0000
@@ -107,7 +107,7 @@
 {
 	void __iomem *vir_addr = NULL;
 
-	vir_addr = ioremap_nocache(addr, 0x100);
+	vir_addr = ioremap(addr, 0x100);
 	if (!vir_addr) {
 		pr_err("ioremap fail\n");
 		return -1;
@@ -125,7 +125,7 @@
 {
 	void __iomem *vir_addr = NULL;
 
-	vir_addr = ioremap_nocache(addr, 0x100);
+	vir_addr = ioremap(addr, 0x100);
 	if (!vir_addr) {
 		pr_err("ioremap fail\n");
 		return -1;
diff --git a/src/conninfra_dev.c b/src/conninfra_dev.c
--- a/src/conninfra_dev.c	2023-12-29 05:52:41.000000000 +0000
+++ b/src/conninfra_dev.c	2026-07-20 13:04:40.620632692 +0000
@@ -240,7 +240,7 @@
 		goto err1;
 	}
 
-	pConninfraClass = class_create(THIS_MODULE, CONNINFRA_DEVICE_NAME);
+	pConninfraClass = class_create(CONNINFRA_DRVIER_NAME);
 	if (IS_ERR(pConninfraClass)) {
 		pr_err("class create fail, error code(%ld)\n", PTR_ERR(pConninfraClass));
 		goto err1;
@@ -341,4 +341,3 @@
 MODULE_AUTHOR("Willy.Yu @ CTD/SE5/CS5");
 
 module_param(gConnInfraMajor, uint, 0644);
-
