--- a/package/network/utils/iwinfo/src/mtwifi.h
+++ b/package/network/utils/iwinfo/src/mtwifi.h
@@ -62,8 +62,8 @@
 #define OID_802_11_BW						0x1903
 #define OID_GET_CHAN_LIST					0x0998
 #define OID_GET_WIRELESS_BAND				0x09B4
-#define OID_GET_CEN_CH1					0x09C0
-#define OID_GET_CEN_CH2					0x09C1
+#define OID_GET_CEN_CH1					0x09F0
+#define OID_GET_CEN_CH2					0x09F1
 #define OID_802_11_SECURITY_TYPE                0x093e
 #define RT_OID_802_11_PHY_MODE				0x050C
 #define GET_MAC_TABLE_STRUCT_FLAG_RAW_SSID	0x1
